linkerd/buffer/src/dispatch.rs

```diff
@@ -84,63 +93,213 @@ where
+                        if this.current_idle.is_none() {
+                            *this.current_idle = Some(delay_for(*timeout));
+                        }
+                    }
+
+                    if let Some(idle) = this.current_idle.as_mut() {
+                        if idle.poll_unpin(cx).is_ready() {
+                            let error = IdleError(*this.max_idle.as_ref().unwrap());
+
+                            // Drop the inner Service to free its resources. It won't be used again.
+                            trace!(%error, "Dropping buffered service");
+                            let _ = this.inner.take();
+
+                            // First, notify services of the readiness change to
+                            // prevent new requests from being buffered. Ensure
+                            // the task remains active until all services have
+                            // observed the error.
+                            let shared = ServiceError(Arc::new(error.into()));
+                            return_ready_if!(this
+                                .ready
+                                .broadcast(Poll::Ready(Err(shared)))
+                                .is_err());
+                        }
+                    }
+
+                    return Poll::Pending;
+                }
+
                 // All senders have been dropped, complete.
-                None => return_ready!(),
+                Poll::Ready(None) => return_ready!(),
 
-                // If a request was ready return it to the caller.
-                Some(InFlight { request, tx }) => {
+                // If a request was ready, dispatch it and return the future to the caller.
+                Poll::Ready(Some(InFlight { request, tx })) => {
                     trace!("Dispatching a request");
                     let fut = this
                         .inner
                         .as_mut()
                         .expect("Service must not be dropped")
                         .call(request);
                     let _ = tx.send(Ok(fut));
+
+                    let _ = this.current_idle.take();
                 }
             }
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use std::task::Poll;
+    use std::time::Duration;
+    use tokio::sync::{mpsc, oneshot, watch};
+    use tokio::time::delay_for;
+    use tokio_test::task;
+    use tower_test::mock;
+
+    #[tokio::test]
+    async fn idle_when_unused() {
+        let max_idle = Duration::from_millis(100);
+
+        let (tx, rx) = mpsc::channel(1);
+        let (ready_tx, ready_rx) = watch::channel(Poll::Pending);
+        let (inner, mut handle) = mock::pair::<(), ()>();
+        let mut dispatch = task::spawn(super::Dispatch::new(inner, rx, ready_tx, Some(max_idle)));
+        handle.allow(1);
+
+        // Service ready without requests. Idle counter starts ticking.
+        assert!(dispatch.poll().is_pending());
+        assert!(ready_rx.borrow().is_ready());
```

Comment:
nit/TIOLI: `tokio_test` also has `assert_ready!` and `assert_pending!` macros...the `assert_ready!` macro has the useful property of also unwrapping the value inside the `Poll::Ready` so you can make additional assertions about it. may or may not be necessary here but they're kinda nice.